In [ ]:
pip install pandas

In [48]:
import pandas as pd

# 1 Extração dos Dados

In [49]:
# Dicionário com os caminhos dos arquivos do ProUni
arquivos = {
    '2017': 'data/pda-prouni-2017.csv',
    '2018': 'data/pda-prouni-2018.csv',
    '2019': 'data/pda-prouni-2019.csv'
}

lista_dataframes = [] # Armazena os DataFrames temporariamente

for ano, arquivo in arquivos.items():
    try:
        df = pd.read_csv(
            arquivo, sep=',', 
            encoding='utf-8-sig',
            on_bad_lines='skip')

        lista_dataframes.append(df)

    except Exception as e:
        print(f"Erro ao ler os dados do ano {ano}: {e}")

df = pd.concat(lista_dataframes, ignore_index=True) #Juntando os DFs
print(f"Total de linhas no df unificado: {len(df)}")

Total de linhas no df unificado: 718700


C:\Users\Clara\AppData\Local\Temp\ipykernel_5552\3618018137.py:12: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,13,14,15,16) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(


# 2 Transformação dos dados

In [50]:
# Analisando o padrão do dataset
display(df.head(5))
display(df.info())

,ANO_CONCESSAO_BOLSA,CODIGO_EMEC_IES_BOLSA,NOME_IES_BOLSA,TIPO_BOLSA,MODALIDADE_ENSINO_BOLSA,NOME_CURSO_BOLSA,NOME_TURNO_CURSO_BOLSA,CPF_BENEFICIARIO_BOLSA,SEXO_BENEFICIARIO_BOLSA,RACA_BENEFICIARIO_BOLSA,DT_NASCIMENTO_BENEFICIARIO,BENEFICIARIO_DEFICIENTE_FISICO,REGIAO_BENEFICIARIO_BOLSA,SIGLA_UF_BENEFICIARIO_BOLSA,MUNICIPIO_BENEFICIARIO_BOLSA,SEMESTRE_CONCESSAO_BOLSA,DT_CONCESSAO_BOLSA
0,2017.0,10.0,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,Presencial,Ciências Contábeis,Noturno,***72056999**,F,Branca,29/04/1992,N,Sul,PR,SAO JOSE DOS PINHAIS,1º Semestre,20-02-2017
1,2017.0,10.0,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,Presencial,Ciências Contábeis,Noturno,***74440980**,M,Branca,24/01/1999,N,Sul,PR,TOLEDO,2º Semestre,26-06-2017
2,2017.0,10.0,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,Presencial,Ciências Contábeis,Noturno,***78689980**,M,Branca,19/05/1994,N,Sul,PR,CURITIBA,1º Semestre,20-02-2017
3,2017.0,10.0,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,Presencial,Ciências Contábeis,Noturno,***90299907**,M,Branca,31/01/1997,N,Sul,PR,TOLEDO,1º Semestre,06-02-2017
4,2017.0,10.0,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,Presencial,Ciências Contábeis,Noturno,***90445902**,M,Branca,24/11/1994,N,Sul,PR,CURITIBA,1º Semestre,06-02-2017


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 718700 entries, 0 to 718699
Data columns (total 17 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   ANO_CONCESSAO_BOLSA             703223 non-null  float64
 1   CODIGO_EMEC_IES_BOLSA           703223 non-null  float64
 2   NOME_IES_BOLSA                  703223 non-null  object 
 3   TIPO_BOLSA                      703223 non-null  object 
 4   MODALIDADE_ENSINO_BOLSA         703223 non-null  object 
 5   NOME_CURSO_BOLSA                703223 non-null  object 
 6   NOME_TURNO_CURSO_BOLSA          703223 non-null  object 
 7   CPF_BENEFICIARIO_BOLSA          703223 non-null  object 
 8   SEXO_BENEFICIARIO_BOLSA         703223 non-null  object 
 9   RACA_BENEFICIARIO_BOLSA         703223 non-null  object 
 10  DT_NASCIMENTO_BENEFICIARIO      703223 non-null  object 
 11  BENEFICIARIO_DEFICIENTE_FISICO  703223 non-null  object 
 12  REGIAO_BENEFICIA

None

### 2.1 Valores ausentes

In [51]:
print("Quantidade de instâncias ausentes por coluna:")
print(df.isnull().sum())

Quantidade de instâncias ausentes por coluna:
ANO_CONCESSAO_BOLSA               15477
CODIGO_EMEC_IES_BOLSA             15477
NOME_IES_BOLSA                    15477
TIPO_BOLSA                        15477
MODALIDADE_ENSINO_BOLSA           15477
NOME_CURSO_BOLSA                  15477
NOME_TURNO_CURSO_BOLSA            15477
CPF_BENEFICIARIO_BOLSA            15477
SEXO_BENEFICIARIO_BOLSA           15477
RACA_BENEFICIARIO_BOLSA           15477
DT_NASCIMENTO_BENEFICIARIO        15477
BENEFICIARIO_DEFICIENTE_FISICO    15477
REGIAO_BENEFICIARIO_BOLSA         15477
SIGLA_UF_BENEFICIARIO_BOLSA       15477
MUNICIPIO_BENEFICIARIO_BOLSA      15477
SEMESTRE_CONCESSAO_BOLSA          15477
DT_CONCESSAO_BOLSA                15477
dtype: int64


Todas as colunas, exceto pelo ano_origem, possuem a mesma quantidade de valores ausentes, ou seja, há chance de que esses valores ausentes ocorram nas mesmas instâncias. Caso essa hipótese esteja correta, basta remover essas instâncias de pouca informação.

In [52]:
# Análise dos valores nulos
mascara_nulos = df.isnull().all(axis=1)

linhas_completamente_nulas = df[mascara_nulos]

print("Linhas que são totalmente nulas:")
print(linhas_completamente_nulas)

Linhas que são totalmente nulas:
        ANO_CONCESSAO_BOLSA  CODIGO_EMEC_IES_BOLSA NOME_IES_BOLSA TIPO_BOLSA  \
703223                  NaN                    NaN            NaN        NaN   
703224                  NaN                    NaN            NaN        NaN   
703225                  NaN                    NaN            NaN        NaN   
703226                  NaN                    NaN            NaN        NaN   
703227                  NaN                    NaN            NaN        NaN   
...                     ...                    ...            ...        ...   
718695                  NaN                    NaN            NaN        NaN   
718696                  NaN                    NaN            NaN        NaN   
718697                  NaN                    NaN            NaN        NaN   
718698                  NaN                    NaN            NaN        NaN   
718699                  NaN                    NaN            NaN        NaN   

      

Para essa análise, a coluna ano_origem foi desconsiderada por ser imputada manualmente ao unificar os datasets. Com isso, a hipótese se mostrou correta e todos os 15477 valores nulos tratam-se das mesmas linhas, as quais devem ser removidas.

In [53]:
# Descrição das colunas para que a remoção das instâncias nulas seja controlada
total_nulas = df[df.columns].isnull().all(axis=1).sum()
print(f"Quantidade de linhas que serão removidas: {total_nulas}")

df.dropna(subset=df.columns, how='all',inplace=True)

Quantidade de linhas que serão removidas: 15477


In [54]:
# Confirmação da remoção das colunas
print(df.isnull().sum())

ANO_CONCESSAO_BOLSA               0
CODIGO_EMEC_IES_BOLSA             0
NOME_IES_BOLSA                    0
TIPO_BOLSA                        0
MODALIDADE_ENSINO_BOLSA           0
NOME_CURSO_BOLSA                  0
NOME_TURNO_CURSO_BOLSA            0
CPF_BENEFICIARIO_BOLSA            0
SEXO_BENEFICIARIO_BOLSA           0
RACA_BENEFICIARIO_BOLSA           0
DT_NASCIMENTO_BENEFICIARIO        0
BENEFICIARIO_DEFICIENTE_FISICO    0
REGIAO_BENEFICIARIO_BOLSA         0
SIGLA_UF_BENEFICIARIO_BOLSA       0
MUNICIPIO_BENEFICIARIO_BOLSA      0
SEMESTRE_CONCESSAO_BOLSA          0
DT_CONCESSAO_BOLSA                0
dtype: int64


### 2.2 Correção do nome dos atributos e dos valores

In [55]:
# Modificando o nome dos atributos para minúsculo
df.columns = df.columns.str.lower()

# Padronização para todos os valores estarem em maiúsculo
for col in df.select_dtypes(include=['object', 'string']).columns:
    df[col] = df[col].astype(str).str.upper()

df.head(5)

,ano_concessao_bolsa,codigo_emec_ies_bolsa,nome_ies_bolsa,tipo_bolsa,modalidade_ensino_bolsa,nome_curso_bolsa,nome_turno_curso_bolsa,cpf_beneficiario_bolsa,sexo_beneficiario_bolsa,raca_beneficiario_bolsa,dt_nascimento_beneficiario,beneficiario_deficiente_fisico,regiao_beneficiario_bolsa,sigla_uf_beneficiario_bolsa,municipio_beneficiario_bolsa,semestre_concessao_bolsa,dt_concessao_bolsa
0,2017.0,10.0,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,PRESENCIAL,CIÊNCIAS CONTÁBEIS,NOTURNO,***72056999**,F,BRANCA,29/04/1992,N,SUL,PR,SAO JOSE DOS PINHAIS,1º SEMESTRE,20-02-2017
1,2017.0,10.0,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,PRESENCIAL,CIÊNCIAS CONTÁBEIS,NOTURNO,***74440980**,M,BRANCA,24/01/1999,N,SUL,PR,TOLEDO,2º SEMESTRE,26-06-2017
2,2017.0,10.0,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,PRESENCIAL,CIÊNCIAS CONTÁBEIS,NOTURNO,***78689980**,M,BRANCA,19/05/1994,N,SUL,PR,CURITIBA,1º SEMESTRE,20-02-2017
3,2017.0,10.0,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,PRESENCIAL,CIÊNCIAS CONTÁBEIS,NOTURNO,***90299907**,M,BRANCA,31/01/1997,N,SUL,PR,TOLEDO,1º SEMESTRE,06-02-2017
4,2017.0,10.0,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,PRESENCIAL,CIÊNCIAS CONTÁBEIS,NOTURNO,***90445902**,M,BRANCA,24/11/1994,N,SUL,PR,CURITIBA,1º SEMESTRE,06-02-2017


### 2.3 Valores únicos

Nessa parte, a análise é sobre quais colunas têm atributos de mesmo valor, mas com escritas diferentes (exemplo: OUTROS e OUTRAS). Além disso, também será analisada a localidade dos dados, se cumpre o requisito do projeto ter dados apenas de Pernambuco.

In [56]:
# Identifica valores únicos
df.nunique(dropna=False)

ano_concessao_bolsa                    3
codigo_emec_ies_bolsa               1485
nome_ies_bolsa                      1469
tipo_bolsa                             2
modalidade_ensino_bolsa                2
nome_curso_bolsa                    2266
nome_turno_curso_bolsa                 5
cpf_beneficiario_bolsa            549249
sexo_beneficiario_bolsa                2
raca_beneficiario_bolsa                6
dt_nascimento_beneficiario         14897
beneficiario_deficiente_fisico         2
regiao_beneficiario_bolsa              5
sigla_uf_beneficiario_bolsa           27
municipio_beneficiario_bolsa        5209
semestre_concessao_bolsa               2
dt_concessao_bolsa                    12
dtype: int64

In [57]:
# Análise de valores únicos
print("Valores únicos da coluna 'Ano de concessão da bolsa': \n", df['ano_concessao_bolsa'].unique())
print()

print("Valores únicos da coluna 'Código EMEC': \n", df['codigo_emec_ies_bolsa'].value_counts().head(5))
print()

print("Valores únicos da coluna 'Nome IES': \n", df['nome_ies_bolsa'].value_counts().head(5))
print()

print("Valores únicos da coluna 'Tipo de bolsa': \n", df['tipo_bolsa'].unique())
print()

print("Valores únicos da coluna 'Modalidade de Ensino Bolsa': \n", df['modalidade_ensino_bolsa'].unique())
print()

print("Valores únicos da coluna 'Nome do curso': \n", df['nome_curso_bolsa'].value_counts().head(5))
print()

print("Valores únicos da coluna 'Turno do curso': \n", df['nome_turno_curso_bolsa'].unique())

Valores únicos da coluna 'Ano de concessão da bolsa': 
 [2017. 2018. 2019.]

Valores únicos da coluna 'Código EMEC': 
 codigo_emec_ies_bolsa
322.0     68539
1491.0    29943
298.0     25296
316.0     14518
671.0     13937
Name: count, dtype: int64

Valores únicos da coluna 'Nome IES': 
 nome_ies_bolsa
UNIVERSIDADE PAULISTA                 68539
CENTRO UNIVERSITÁRIO INTERNACIONAL    29943
UNIVERSIDADE PITÁGORAS UNOPAR         25296
UNIVERSIDADE NOVE DE JULHO            14518
UNIVERSIDADE ANHANGUERA               13937
Name: count, dtype: int64

Valores únicos da coluna 'Tipo de bolsa': 
 ['BOLSA PARCIAL 50%' 'BOLSA INTEGRAL']

Valores únicos da coluna 'Modalidade de Ensino Bolsa': 
 ['PRESENCIAL' 'EAD']

Valores únicos da coluna 'Nome do curso': 
 nome_curso_bolsa
ADMINISTRAÇÃO         53285
DIREITO               49916
PEDAGOGIA             48004
CIÊNCIAS CONTÁBEIS    32462
ENFERMAGEM            27549
Name: count, dtype: int64

Valores únicos da coluna 'Turno do curso': 
 ['NOTURNO' 'MAT

In [58]:
# Divisão em blocos para melhor visualização
print("Valores únicos da coluna 'CPF do beneficiário': \n", df['cpf_beneficiario_bolsa'].value_counts().head(5))
print()

print("Valores unicos da coluna 'Sexo do beneficiário': \n", df['sexo_beneficiario_bolsa'].unique())
print()

print("Valores únicos da coluna 'Raça/Cor do beneficiário': \n", df['raca_beneficiario_bolsa'].unique())
print()

print("Valores únicos da coluna 'Data de nascimento do beneficiário': \n", df['dt_nascimento_beneficiario'].unique())
print()

print("Valores únicos da coluna 'Beneficiário é deficiente físico': \n", df['beneficiario_deficiente_fisico'].unique())
print()

print("Valores unicos da coluna 'Região do beneficiário': \n", df['regiao_beneficiario_bolsa'].unique())
print()

print("Valores únicos da coluna 'Sigla UF do beneficiário': \n", df['sigla_uf_beneficiario_bolsa'].value_counts().head(5))
print()

print("Valores unicos da coluna 'Município do beneficiário': \n", df['municipio_beneficiario_bolsa'].value_counts().head(5))

Valores únicos da coluna 'CPF do beneficiário': 
 cpf_beneficiario_bolsa
***21978486**    12
***55193851**    12
***12884025**    12
***90782040**    12
***47410505**    11
Name: count, dtype: int64

Valores unicos da coluna 'Sexo do beneficiário': 
 ['F' 'M']

Valores únicos da coluna 'Raça/Cor do beneficiário': 
 ['BRANCA' 'PARDA' 'PRETA' 'AMARELA' 'INDÍGENA' 'NÃO INFORMADA']

Valores únicos da coluna 'Data de nascimento do beneficiário': 
 ['29/04/1992' '24/01/1999' '19/05/1994' ... '11/04/1969' '28/12/1954'
 '24/02/1963']

Valores únicos da coluna 'Beneficiário é deficiente físico': 
 ['N' 'S']

Valores unicos da coluna 'Região do beneficiário': 
 ['SUL' 'SUDESTE' 'NORTE' 'NORDESTE' 'CENTRO-OESTE']

Valores únicos da coluna 'Sigla UF do beneficiário': 
 sigla_uf_beneficiario_bolsa
SP    184313
MG     81097
RS     49654
PR     48207
BA     42989
Name: count, dtype: int64

Valores unicos da coluna 'Município do beneficiário': 
 municipio_beneficiario_bolsa
SAO PAULO         53169
BRA

In [59]:
print("Valores únicos da coluna 'Data de concessão da bolsa': \n", df['dt_concessao_bolsa'].value_counts().head(12))
print()

print("Valores unicos da coluna 'Semestre de concessão da bolsa': \n", df['semestre_concessao_bolsa'].unique())
print()

Valores únicos da coluna 'Data de concessão da bolsa': 
 dt_concessao_bolsa
14-02-2018    60364
02-03-2018    60322
16-07-2018    60178
02-07-2018    60168
26-06-2017    59382
12-06-2017    59281
06-02-2017    59061
20-02-2017    58912
06-02-2019    56453
20-02-2019    56410
02-07-2019    56399
18-06-2019    56293
Name: count, dtype: int64

Valores unicos da coluna 'Semestre de concessão da bolsa': 
 ['1º SEMESTRE' '2º SEMESTRE']



A análise dos valores repetidos de CPF não indica que necessariamente há repetição de participantes. Como o número do CPF é parcialmente ocultado, é possível que a sequência do meio seja igual, mas com diferença no início ou no final. Não foi identificada nenhum desvio de escrita entre os datasets que exigisse alteração.

### 2.3 Correção de tipagem de atributos

Com base nas visualizações da etapa anterior, é possível classificar corretamente o tipo de cada atributo e fazer a conversão necessária.

In [60]:
# São as duas únicas colunas cujo tipo é float, estão sendo convertidas para inteiros
df['ano_concessao_bolsa'] = pd.to_numeric(df['ano_concessao_bolsa'], errors='coerce').astype('Int64')
df['codigo_emec_ies_bolsa'] = pd.to_numeric(df['codigo_emec_ies_bolsa'], errors='coerce').astype('Int64')

In [61]:
# Conversão para string
df['nome_ies_bolsa'] = df['nome_ies_bolsa'].astype(str)
df['tipo_bolsa'] = df['tipo_bolsa'].astype(str)
df['modalidade_ensino_bolsa'] = df['modalidade_ensino_bolsa'].astype(str)
df['nome_curso_bolsa'] = df['nome_curso_bolsa'].astype(str)
df['nome_turno_curso_bolsa'] = df['nome_turno_curso_bolsa'].astype(str)
df['raca_beneficiario_bolsa'] = df['raca_beneficiario_bolsa'].astype(str)
df['regiao_beneficiario_bolsa'] = df['regiao_beneficiario_bolsa'].astype(str)
df['municipio_beneficiario_bolsa'] = df['municipio_beneficiario_bolsa'].astype(str)
df['sigla_uf_beneficiario_bolsa'] = df['sigla_uf_beneficiario_bolsa'].astype(str)
df['cpf_beneficiario_bolsa'] = df['cpf_beneficiario_bolsa'].astype(str)
df['sexo_beneficiario_bolsa'] = df['sexo_beneficiario_bolsa'].astype(str)
df['beneficiario_deficiente_fisico'] = df['beneficiario_deficiente_fisico'].astype(str)
df['semestre_concessao_bolsa'] = df['semestre_concessao_bolsa'].astype(str)

In [81]:
# É importante ressaltar que o formato utilizado é DD/MM/YYYY
df['dt_nascimento_beneficiario'] = pd.to_datetime(df['dt_nascimento_beneficiario'], format='%d/%m/%Y', errors='coerce')
df['dt_concessao_bolsa'] = pd.to_datetime(df['dt_concessao_bolsa'], format='%d-%m-%Y', errors='coerce')

In [82]:
# Visualização do resultado da transformação
display(df.info())
display(df.head())

<class 'pandas.core.frame.DataFrame'>
Index: 703223 entries, 0 to 703222
Data columns (total 17 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   ano_concessao_bolsa             703223 non-null  Int64         
 1   codigo_emec_ies_bolsa           703223 non-null  Int64         
 2   nome_ies_bolsa                  703223 non-null  object        
 3   tipo_bolsa                      703223 non-null  object        
 4   modalidade_ensino_bolsa         703223 non-null  object        
 5   nome_curso_bolsa                703223 non-null  object        
 6   nome_turno_curso_bolsa          703223 non-null  object        
 7   cpf_beneficiario_bolsa          703223 non-null  object        
 8   sexo_beneficiario_bolsa         703223 non-null  object        
 9   raca_beneficiario_bolsa         703223 non-null  object        
 10  dt_nascimento_beneficiario      0 non-null       datetime64[n

None

,ano_concessao_bolsa,codigo_emec_ies_bolsa,nome_ies_bolsa,tipo_bolsa,modalidade_ensino_bolsa,nome_curso_bolsa,nome_turno_curso_bolsa,cpf_beneficiario_bolsa,sexo_beneficiario_bolsa,raca_beneficiario_bolsa,dt_nascimento_beneficiario,beneficiario_deficiente_fisico,regiao_beneficiario_bolsa,sigla_uf_beneficiario_bolsa,municipio_beneficiario_bolsa,semestre_concessao_bolsa,dt_concessao_bolsa
0,2017,10,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,PRESENCIAL,CIÊNCIAS CONTÁBEIS,NOTURNO,***72056999**,F,BRANCA,NaT,N,SUL,PR,SAO JOSE DOS PINHAIS,1º SEMESTRE,2017-02-20
1,2017,10,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,PRESENCIAL,CIÊNCIAS CONTÁBEIS,NOTURNO,***74440980**,M,BRANCA,NaT,N,SUL,PR,TOLEDO,2º SEMESTRE,2017-06-26
2,2017,10,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,PRESENCIAL,CIÊNCIAS CONTÁBEIS,NOTURNO,***78689980**,M,BRANCA,NaT,N,SUL,PR,CURITIBA,1º SEMESTRE,2017-02-20
3,2017,10,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,PRESENCIAL,CIÊNCIAS CONTÁBEIS,NOTURNO,***90299907**,M,BRANCA,NaT,N,SUL,PR,TOLEDO,1º SEMESTRE,2017-02-06
4,2017,10,PONTIFÍCIA UNIVERSIDADE CATÓLICA DO PARANÁ,BOLSA PARCIAL 50%,PRESENCIAL,CIÊNCIAS CONTÁBEIS,NOTURNO,***90445902**,M,BRANCA,NaT,N,SUL,PR,CURITIBA,1º SEMESTRE,2017-02-06


# 3 Carregamento e transformação em esquema estrela

In [64]:
# Instalando as dependências necessárias
!pip install sqlalchemy
!pip install psycopg2-binary

^C



### 3.1 Criação do esquema estrela

Para que nosso banco de dados seja transformado em esquema estrela, é necessário dividir seus atributos de forma que hajam dimensões e a tabela fato.

As seguintes dimensões serão criadas:
* Dimensão beneficiário (cpf, sexo, raça, data de nascimento, presença de deficiência)
* Dimensão localidade (região, sigla da Unidade Federativa, município)
* Dimensão Instituição de Ensino (código EMEC, nome da IES)
* Dimensão curso (nome do curso, turno do curso, modalidade de ensino)
* Dimensão temporal (data de concessão da bolsa, semestre de concessão da bolsa)

In [65]:
# Criando dimensão beneficiário
colunas_beneficiario = [
    'cpf_beneficiario_bolsa',
    'sexo_beneficiario_bolsa',
    'raca_beneficiario_bolsa',
    'dt_nascimento_beneficiario',
    'beneficiario_deficiente_fisico'
    ]

dim_beneficiario = df[colunas_beneficiario].drop_duplicates().reset_index(drop=True)
dim_beneficiario = dim_beneficiario.reset_index().rename(columns={'index': 'id_beneficiario_sk'})

print(f"Dimensão Beneficiário: {len(dim_beneficiario)} perfis")

Dimensão Beneficiário: 553160 perfis


In [66]:
# Criando dimensão localidade
colunas_localidade = [
    'regiao_beneficiario_bolsa',
    'sigla_uf_beneficiario_bolsa',
    'municipio_beneficiario_bolsa'
    ]

dim_localidade = df[colunas_localidade].drop_duplicates().reset_index(drop=True)
dim_localidade = dim_localidade.reset_index().rename(columns={'index': 'id_localidade_sk'})

print(f"Dimensão Localidade: {len(dim_localidade)} perfis")

Dimensão Localidade: 5482 perfis


In [67]:
# Criando dimensão Instituição de Ensino
colunas_ies = [
    'codigo_emec_ies_bolsa',
    'nome_ies_bolsa'
    ]

dim_ies = df[colunas_ies].drop_duplicates().reset_index(drop=True)
dim_ies = dim_ies.reset_index().rename(columns={'index': 'id_ies_sk'})

print(f"Dimensão Instituição de Ensino: {len(dim_ies)} perfis")

Dimensão Instituição de Ensino: 1485 perfis


In [68]:
# Criando dimensão curso
colunas_curso = [
    'nome_curso_bolsa',
    'nome_turno_curso_bolsa',
    'modalidade_ensino_bolsa'
    ]

dim_curso = df[colunas_curso].drop_duplicates().reset_index(drop=True)
dim_curso = dim_curso.reset_index().rename(columns={'index': 'id_curso_sk'})

print(f"Dimensão Curso: {len(dim_curso)} perfis")

Dimensão Curso: 3380 perfis


In [83]:
# Criando dimensão temporal
colunas_interesse = ['dt_concessao_bolsa', 'semestre_concessao_bolsa']
datas_unicas = df[colunas_interesse].drop_duplicates().sort_values('dt_concessao_bolsa').reset_index(drop=True)

# Derivação dos atributos temporais
datas_unicas['dia'] = datas_unicas['dt_concessao_bolsa'].dt.day
datas_unicas['mes'] = datas_unicas['dt_concessao_bolsa'].dt.month
datas_unicas['ano'] = datas_unicas['dt_concessao_bolsa'].dt.year

# Consolidação
dim_temporal = datas_unicas.reset_index().rename(columns={'index': 'id_temporal_sk'})

print(f"Dimensão Temporal criada: {len(dim_temporal)} dias registrados.")

Dimensão Temporal criada: 12 dias registrados.


In [ ]:
# Aplicando a merge para construir a tabela fato
fato = df.merge(dim_beneficiario, on=colunas_beneficiario, how='left')
fato = fato.merge(dim_localidade, on=colunas_localidade, how='left')
fato = fato.merge(dim_ies, on=colunas_ies, how='left')
fato = fato.merge(dim_curso, on=colunas_curso, how='left')
fato = fato.merge(dim_temporal[['id_temporal_sk', 'dt_concessao_bolsa']], 
    on='dt_concessao_bolsa', 
    how='left')

# Aplicando os ids
colunas_fato = [
    'id_beneficiario_sk',
    'id_localidade_sk',
    'id_ies_sk',
    'id_curso_sk',
    'id_temporal_sk',
    'ano_concessao_bolsa',
    'tipo_bolsa',
]

fato_prouni = fato[colunas_fato]
print(f"Fato pronta: {len(fato_prouni)} registros.")

Fato pronta: 703223 registros.


### 3.2 Carregamento

In [72]:
from sqlalchemy import create_engine

In [ ]:
# Informações para conexão
usuario_db = 'postgres'
senha_db = 'senha' # mudar para a senha do usuário quando for rodar
host_db = 'localhost'
porta_db = '5432'
nome_banco_db = 'postgres'

In [74]:
# Criando a "connection string"
DATABASE_URL = f'postgresql://{usuario_db}:{senha_db}@{host_db}:{porta_db}/{nome_banco_db}?client_encoding=utf8'
print("Tentando criar conexão com o banco...")

try:
    engine = create_engine(DATABASE_URL)
    print("Conexão bem-sucedida! Iniciando a carga do Esquema Estrela...")

    print("Carregando Dimensão Beneficiário...")
    dim_beneficiario.to_sql('dim_beneficiario', con=engine, if_exists='replace', index=False)

    print("Carregando Dimensão Localidade...")
    dim_localidade.to_sql('dim_localidade', con=engine, if_exists='replace', index=False)

    print("Carregando Dimensão Instituição de Ensino...")
    dim_ies.to_sql('dim_ies', con=engine, if_exists='replace', index=False)

    print("Carregando Dimensão Curso...")
    dim_curso.to_sql('dim_curso', con=engine, if_exists='replace', index=False)

    print("Carregando Dimensão Temporal...")
    dim_temporal.to_sql('dim_temporal', con=engine, if_exists='replace', index=False)

    print("Carregando Tabela Fato...")
    fato_prouni.to_sql('fato_prouni', con=engine, if_exists='replace', index=False)

    print("\nCARGA CONCLUÍDA! Todas as 6 tabelas do Esquema Estrela foram criadas.")

except Exception as e:
    print(f"\n--- ERRO! ---")
    print(f"Erro ao carregar os dados: {e}")

Tentando criar conexão com o banco...
Conexão bem-sucedida! Iniciando a carga do Esquema Estrela...
Carregando Dimensão Beneficiário...
Carregando Dimensão Localidade...
Carregando Dimensão Instituição de Ensino...
Carregando Dimensão Curso...
Carregando Dimensão Temporal...
Carregando Tabela Fato...

CARGA CONCLUÍDA! Todas as 6 tabelas do Esquema Estrela foram criadas.
